src/protocol.rs

```diff
@@ -0,0 +1,774 @@
+#![allow(dead_code)]
+
+use mockall_double::double;
+
+#[double]
+use crate::targets::esp32c3 as target;
+
+#[derive(Debug, PartialEq)]
+pub enum ErrorIO {
+    Overflow,
+    Incomplete,
+    InvalidResponse,
+    Hardware,
+}
+
+pub trait InputIO {
+    fn recv(&mut self) -> Result<u8, ErrorIO>;
+    fn send(&mut self, data: &[u8]) -> Result<(), ErrorIO>;
+}
+
+use target::*;
+use slip::*;
+use core::mem;
+use core::slice;
+use core::cmp::min;
+use crate::commands::*;
+use crate::commands::CommandCode::*;
+use crate::commands::Error::*;
+use md5::{Md5, Digest};
+use crate::miniz_types::*;
+
+const DATA_CMD_SIZE: usize = mem::size_of::<DataCommand>();
+const CMD_BASE_SIZE: usize = mem::size_of::<CommandBase>();
+
+const FLASH_SECTOR_SIZE: u32 = 4096;
+const MAX_WRITE_BLOCK: u32 = 0x4000;
+pub struct Stub<'a> {
+    io: &'a mut (dyn InputIO + 'a),
+    end_addr: u32,
+    write_addr: u32,
+    erase_addr: u32,
+    remaining: u32,
+    remaining_compressed: usize,
+    decompressor: tinfl_decompressor,
+    last_error: Option<Error>,
+    in_flash_mode: bool,
+}
+
+impl From<ErrorIO> for Error {
+    fn from(_: ErrorIO) -> Self {
+        Error::FailedSpiOp
+    }
+}
+
+// todo: get rid of SIZE
+fn slice_to_struct<T: Sized, const SIZE: usize>(slice: &[u8]) -> Result<T, Error>
+{
```

Comment:
Alignment should be checked too, but since all structs passed into here are packed, their alignment is always 1.